third_party/2and3/redis/client.pyi

```diff
@@ -170,40 +170,51 @@ class Redis(object):
     def blpop(self, keys: Union[_Value, Iterable[_Value]], timeout: int = ...) -> Optional[Tuple[bytes, bytes]]: ...
     def brpop(self, keys: Union[_Value, Iterable[_Value]], timeout: int = ...) -> Optional[Tuple[bytes, bytes]]: ...
     def brpoplpush(self, src, dst, timeout=...): ...
-    def lindex(self, name, index): ...
-    def linsert(self, name, where, refvalue, value): ...
-    def llen(self, name): ...
+    def lindex(self, name: _Key, index: int) -> Optional[bytes]: ...
+    def linsert(self, name: _Key, where: Text, refvalue: _Value, value: _Value) -> int: ...
+    def llen(self, name: _Key) -> int: ...
     def lpop(self, name): ...
     def lpush(self, name: _Value, *values: _Value) -> int: ...
     def lpushx(self, name, value): ...
-    def lrange(self, name, start, end): ...
-    def lrem(self, name, count, value): ...
-    def lset(self, name, index, value): ...
-    def ltrim(self, name, start, end): ...
+    def lrange(self, name: _Key, start: int, end: int) -> List[bytes]: ...
+    def lrem(self, name: _Key, count: int, value: _Value) -> int: ...
+    def lset(self, name: _Key, index: int, value: _Value) -> bool: ...
+    def ltrim(self, name: _Key, start: int, end: int) -> bool: ...
     def rpop(self, name): ...
     def rpoplpush(self, src, dst): ...
     def rpush(self, name: _Value, *values: _Value) -> int: ...
     def rpushx(self, name, value): ...
-    def sort(self, name, start=..., num=..., by=..., get=..., desc=..., alpha=..., store=..., groups=...): ...
+    def sort(
+        self,
+        name: _Key,
+        start: Optional[int] = ...,
+        num: Optional[int] = ...,
+        by: Optional[_Key] = ...,
+        get: Optional[Union[_Key, Sequence[_Key]]] = ...,
+        desc: bool = ...,
+        alpha: bool = ...,
+        store: Optional[_Key] = ...,
+        groups: Optional[bool] = ...,
+    ) -> Union[List[bytes], int]: ...
```

Comment:
Since the result type depends on whether `store` is given, we should use overloads to choose the right type (most arguments omitted in the example for brevity):

```python
@overload
def sort(
    self,
    name: _Key, 
    start: Optional[int] = ...,
    # etc.
    store: None = ...,
    groups: Optional[bool] = ...,
) -> List[bytes]: ...
@overload
def sort(
    self,
    name: _Key, 
    start: Optional[int] = ...,
    # etc. with default
    *,
    store: _Key,
    groups: Optional[bool] = ...,
) -> int: ...
@overload
def sort(
    self,
    name: _Key, 
    start: Optional[int],
    # etc. without default
    store: _Key,
    groups: Optional[bool] = ...,
) -> int: ...
```

Unfortunately we need to overloads for the second case, otherwise we'd have a syntax error.

Reply:
Great catch. However, I'm a little stuck. Whenever I try to implement this solution, I get the following errors:

    stubs/redis/client.pyi:190: error: Overloaded function signatures 1 and 2 overlap with incompatible return types
    stubs/redis/client.pyi:190: error: Overloaded function signatures 1 and 3 overlap with incompatible return types

Any tips?

Reply:
Hmm, I can't reproduce this here locally, although we sometimes need `# type: ignore` when overloading. Could you commit what you have so far? We squash merge anyway.